### ISSUE TITLE
Does not meet the 1.0 spec for removing related records

### ISSUE THREAD

**Issue by geoidesic** [FriendsOfCake/crud-json-api#117]
As per the spec: https://jsonapi.org/format/#crud-updating-to-many-relationships
This should allow one to remove all linked records from a resource using a request like this:
```
PATCH /restuarants/1/relationships/dishes HTTP/1.1
Content-Type: application/vnd.api+json
Accept: application/vnd.api+json

{
  "data": []
}
```
`crud-json-api` does not recognise this URL format. It's possible to force it to like so:
```
    $routes->connect('/{relation}/{id}/relationships/dishes', ['_method' => 'PATCH', 'controller' => 'Dishes', 'action' => 'edit']);
```
but then it produces an error because this part of the spec isn't handled:
```
{
    "errors": [
        {
            "title": "_required",
            "detail": "Primary data does not contain member 'type'",
            "source": {
                "pointer": "/data"
            }
        },
        {
            "title": "_required",
            "detail": "Primary data does not contain member 'id'",
            "source": {
                "pointer": "/data"
            }
        }
    ],
    "query": {
        "default": [],
        "test": [],
        "debug_kit": []
    }
}
```

**Comment by geoidesic** [comment]
I'm looking into how to fix this. The main problem is that the design of DocumentValidator seems to assume that the structure of the data is always the same for every type of JSONAPI request, and thus it does not take the type of request into account. However this assumption is false – the data structure can vary according to the URL pattern, as shown above.

There are two possible routes to fixing this:
1. Pass the request object into DocumentValidator but this will complicate writing tests.
2. Create a second, alternate DocumentValidator (perhaps DocumentRelationshipValidator, which is for this unusual format, then do URL regex matching in JsonApiListener in order to instantiate that validator instead, as appropriate.

I think option two makes the most sense. I would appreciate feedback on this before I begin.

**Comment by geoidesic** [comment]
This brings into question the claim of immutability made here:
https://github.com/FriendsOfCake/crud-json-api/blob/382907f1d9cca4fe23fc9113d93bb9ea45037806/src/Listener/JsonApiListener.php#L992

**Comment by dakota** [comment]
I definately agree that this part of the spec is not currently handled. There is a test to ensure that the route is generated correctly (See https://github.com/FriendsOfCake/crud-json-api/blob/cake-4.x/tests/test_app/config/routes.php#L16-L24)

I'm not sure how this would invalidate the claim of immutability?

**Comment by geoidesic** [comment]
What I understood from the the claim of immutability (and please correct me if I'm wrong) is that it means that the "id" property must be present within the main "data" body and must match that in the URL.

However (as per my first quoted example above) in this case the "id" would be present in the URL but not need to be present in the data body.

Possibly it would be clearer for me to say that while it doesn't invalidate the immutability of the "id" property of the entity itself, the discussion around immutability from the referenced GitHub issue seems to have implied as its outcome the assumption that the "id" property must always be present in the data body... or at least that's how I read it.

**Comment by geoidesic** [comment]
LOL, I'm not sure that was any clearer. Let me try again...

The comment I linked above suggests that the linked GitHub issue (which discusses immutability) is the reason that the id property in the data body must be present and must match that in the URL. I'm not sure whether the GitHub issue really establishes that fact or not because (frankly) TLDR;

However this part of the spec contra-indicates that comment's assertion.

**Comment by dakota** [comment]
Ah, I think you are misunderstanding the purpose of this.

The exact part of the spec that deals with it is https://jsonapi.org/format/1.1/#crud-updating-responses-409 which says

> A server MUST return 409 Conflict when processing a PATCH request in which the resource object’s type and id do not match the server’s endpoint.

Currently don't return a 409, so that's definately something that must be fixed. What is also needed is a check that the type matches the URL, but I'm not sure how that would practically work.

As for the specific case you are referring to here (namely updating a to-many relationship), this above linked section, the comment in the code, and the github issue linked do not apply, as they are specifically referring to updating of a resource.

**Comment by geoidesic** [comment]
Perhaps this is also related? https://github.com/FriendsOfCake/crud-json-api/blob/382907f1d9cca4fe23fc9113d93bb9ea45037806/src/Listener/JsonApiListener.php#L1035

**Comment by geoidesic** [comment]
Recent commits do not address this problem. Sending a PATCH relationships request with an empty data body results in the following error:

```
2020-09-05 16:08:28 Error: [TypeError] Return value of CrudJsonApi\Listener\JsonApiListener::_convertJsonApiDocumentArray() must be of the type array, object returned in /Users/me/vendor/friendsofcake/crud-json-api/src/Listener/JsonApiListener.php on line 1288
Stack Trace:
- /Users/me/vendor/friendsofcake/crud-json-api/src/Listener/JsonApiListener.php:1196
- /Users/me/vendor/friendsofcake/crud-json-api/src/Listener/JsonApiListener.php:128
- /Users/me/vendor/cakephp/cakephp/src/Event/EventManager.php:309
- /Users/me/vendor/cakephp/cakephp/src/Event/EventManager.php:286
- /Users/me/vendor/friendsofcake/crud/src/Controller/Component/CrudComponent.php:550
- /Users/me/vendor/friendsofcake/crud/src/Controller/Component/CrudComponent.php:242
- /Users/me/vendor/friendsofcake/crud/src/Controller/ControllerTrait.php:68
- /Users/me/vendor/cakephp/cakephp/src/Controller/ControllerFactory.php:81
- /Users/me/vendor/cakephp/cakephp/src/Http/BaseApplication.php:251
- /Users/me/vendor/cakephp/cakephp/src/Http/Runner.php:77
- /Users/me/vendor/cakephp/cakephp/src/Http/Runner.php:77
- /Users/me/vendor/cakephp/cakephp/src/Http/Middleware/BodyParserMiddleware.php:174
- /Users/me/vendor/cakephp/cakephp/src/Http/Runner.php:73
- /Users/me/plugins/Cors/src/Middleware/CorsMiddleware.php:39
- /Users/me/vendor/cakephp/cakephp/src/Http/Runner.php:73
- /Users/me/vendor/cakephp/cakephp/src/Http/Runner.php:58
- /Users/me/vendor/cakephp/cakephp/src/Routing/Middleware/RoutingMiddleware.php:172
- /Users/me/vendor/cakephp/cakephp/src/Http/Runner.php:73
- /Users/me/vendor/cakephp/cakephp/src/Routing/Middleware/AssetMiddleware.php:68
- /Users/me/vendor/cakephp/cakephp/src/Http/Runner.php:73
- /Users/me/vendor/cakephp/cakephp/src/Error/Middleware/ErrorHandlerMiddleware.php:121
- /Users/me/vendor/cakephp/cakephp/src/Http/Runner.php:73
- /Users/me/vendor/cakephp/cakephp/src/Http/Runner.php:58
- /Users/me/vendor/cakephp/cakephp/src/Http/Server.php:90
- /Users/me/webroot/index.php:40

Request URL: /api/restaurants/38/relationships/dishes
Referer URL: http://localhost:8000
Client IP: 127.0.0.1
```

**Comment by geoidesic** [comment]
Fixed in current version of `cake-4.x` branch.
